방 목록 검색 조건 수정 및 실시간 갱신 기능 구현

yxxjxxeee opened the pull request:
## #️⃣ 연관된 이슈
- #32 

## 📝 작업 내용
> 이번 PR에서는 방 목록 관련 기능 개선과 실시간 갱신 기능을 구현했습니다. 
> 1. 상수명 네이밍 컨벤션 통일 (일관된 코드 스타일 유지)
> 2. 방 목록 검색 조건 수정 (필터 및 정렬 조건 개선)
> 3. Socket.io 이벤트 도입을 통한 방 생성, 수정, 삭제 실시간 갱신

## 💬 리뷰 요구사항 (선택)
> RoomMember 기반 기능(방장 권한 체크, 참여자 0명 시 방 삭제, 방장 자동 인계 등)은 다음 PR에서 구현할 예정입니다.

yxxjxxeee committed:
refactor: (#32) 에러 상수에 ROOM 접두사를 붙여 범위 명확화

yxxjxxeee committed:
refactor: (#32) 방 목록 검색 조건 개선

yxxjxxeee committed:
Merge branch 'develop' of https://github.com/modern-agile-team/10term-pollycasso-back into feat/#32/rooms-socketio

yxxjxxeee committed:
git commit -m "Merge branch into feat/#32/rooms-socketio"

yxxjxxeee committed:
feat: (#32) 방 생성, 수정, 삭제 실시간 기능 구현

yxxjxxeee committed:
refactor: (#32) swagger 응답 코드 변경

rrgks6221 commented on src/rooms/dtos/requests/query-room.dto.ts:
```diff
@@ -1,16 +1,16 @@
     required: false,
   })
-  name?: string;
+  keyword?: string;
```
지금도 좋습니다만 검색을 위한 query parameter로는 q를 자주 사용하곤 합니다

rrgks6221 commented on src/rooms/rooms.gateway.ts:
```diff
@@ -0,0 +1,28 @@
+  },
+  namespace: '/rooms',
+})
+export class RoomsGateway {
```
게이트웨이는 입출력 포트(controller와 비슷한 역할)을 합니다.
service를 보니 RoomGateWay를 그대로 DI 받는것으로 확인했어요.
service가 controller를 DI 받는다면 어색하다고 느껴져요.

당장에 socket server에 접근하기 위해서 WebSocketGateWay 데코레이터를 사용하니 클래스 명 자체는 RoomsGateway로 하는것에는 동의하는데요.
이벤트 퍼블리싱만을 위한 port interface를 따로 두어 하는건 어떻게 생각하실까요??
지금처럼 구현체 자체를 의존한다면 불필요한 인터페이스까지 접근이 가능해집니다.(DIP)

이벤트 퍼블리셔만 똑 떼는게 가장 좋긴 할텐데 그려려면 다른 작업들이 많아질거같아서 인테페이스만 분리되면 좋을거같습니다

```suggestion
export class RoomsGateway implements IRoomsEventPublisher {

export interface IRoomsEventPublisher {
  roomCreated(room: Room): void
  //... 
}
```

rrgks6221 replied:
혹시 문서화가 필요하시다면 nestjs-asyncapi 가 생각보다 잘 해주더라고요

yxxjxxeee replied:
안그래도 노션에 정리하려고 했는데 이런 게 있군요 
알려주셔서 감사해요!

rrgks6221 commented on src/rooms/rooms.repository.ts:
```diff
@@ -58,10 +58,21 @@ export class RoomsRepository implements IRoomsRepository {
+      if (isIdSearch) {
+        where.id = Number(keyword);
+      } else {
+        where.name = { contains: keyword };
```
어플리케이션 배포 후 시간이 남으신다면 검색을 위한 인덱싱도 공부해보시고 적용해보시면 성장에 도움이 될거같아요. (일반적인 b-tree 인덱싱이 아닌 full text search 인덱싱)

일반적으로 게임방은 개수가 적으니 성능면에서는 눈에 보일만한 차이는 없을거같아요

yxxjxxeee replied:
좋은 조언 감사합니다! 
시간 날 때 공부해서 적용해보겠습니다.

yxxjxxeee committed:
refactor: (#32) 분리된 IRoomsEventPublisher 인터페이스 적용

- Service가 Gateway 구현체에 직접 의존하지 않고 DIP 원칙 준수
- nestjs-asyncapi 기반 WebSocket 이벤트 문서화

yxxjxxeee committed:
refactor: (#32) roomGateway CORS origin 설정 리팩토링

yxxjxxeee committed:
refactor: (#32) CORS origin 처리 방식 수정
